segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ThrottlerCalculator.java

```diff
@@ -0,0 +1,217 @@
+
+    //endregion
+
+    //region Methods
+
+    /**
+     * Determines whether there is an immediate need to introduce a throttling delay.
+     *
+     * @return True if throttling is required, false otherwise.
+     */
+    boolean isThrottlingRequired() {
+        for (Throttler t : this.throttlers) {
+            if (t.isThrottlingRequired()) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Calculates the amount of time needed to delay based on the configured throttlers. The computed result is not additive,
+     * as there is no benefit to adding delays from various Throttle Calculators together. For example, a cache throttling
+     * delay will have increased batching as a side effect, so there's no need to include the batching one as well.
+     *
+     * @return A DelayResult representing the computed delay. This delay is the maximum delay as calculated by the internal
+     * throttlers.
+     */
+    DelayResult getThrottlingDelay() {
+        // These delays are not additive. There's no benefit to adding a batching delay on top of a throttling delay, since
+        // a throttling delay will have increased batching as a side effect.
+        int maxDelay = 0;
+        boolean maximum = false;
+        for (Throttler t : this.throttlers) {
+            DelayResult r = t.getDelay();
+            maxDelay = Math.max(maxDelay, r.durationMillis);
+            maximum |= r.maximum;
+        }
+
+        return new DelayResult(maxDelay, maximum);
+    }
+
+    //endregion
+
+    //region Throttlers
+
+    /**
+     * Defines the structure of a Throttle Calculator.
+     */
+    static abstract class Throttler {
+        /**
+         * Determines whether throttling based on this calculator is absolutely required at this moment.
+         */
+        abstract boolean isThrottlingRequired();
+
+        /**
+         * Calculates a throttling delay based on information available at the moment.
+         */
+        abstract DelayResult getDelay();
+    }
+
+    /**
+     * Calculates the amount of time to wait before processing more operations from the queue in order to relieve pressure
+     * on the cache. This is based on ReadIndex statistics, mainly cache utilization ratio.
+     */
+    @RequiredArgsConstructor
+    private static class CacheThrottler extends Throttler {
+        private final Supplier<Double> getCacheUtilization;
+
+        @Override
+        boolean isThrottlingRequired() {
+            return this.getCacheUtilization.get() > 1;
+        }
+
+        @Override
+        DelayResult getDelay() {
+            // We only throttle if we exceed the cache capacity. We increase the throttling amount in a linear fashion.
+            double cacheUtilization = this.getCacheUtilization.get();
+            int delay = (int) MathHelpers.minMax((cacheUtilization - 1.0) * 100 * THROTTLING_MILLIS_PER_PERCENT_OVER_LIMIT, 0, MAX_DELAY_MILLIS);
```

Comment:
It does work that way, but not by looking at just this data. If our situation gets worse and worse, eventually we'll be returning our maximum allowed value (we cap at that constant). If that's the case, then we set that flag (`isMaximumLimit`) which the `OperationProcessor` will check that and continue throttling as long as the flag is set (check the `OperationProcessor.throttle` method for that).

Reply:
This formula:

```
int delay = (int) MathHelpers.minMax((cacheUtilization - 1.0) * 100 * THROTTLING_MILLIS_PER_PERCENT_OVER_LIMIT, 0, MAX_DELAY_MILLIS);
```

is purely based on cache utilization, no? I guess I'm missing something.

Reply:
Yes. This is just a calculator. This value is used in the OperationProcessor which does the actual throttling work. 

If you want to consider "recent memory", then you can think of the Cache itself as holding that information. The Cache Manager's actions leave the Cache in the best shape it can; it's trying hard to keep it under the limit. If it increases beyond the limit, then we throttle a little bit. If the situation doesn't change, then we reached an equilibrium point, which is fine. If the situation worsens, then we will throttle more, and so on until we either reach a new equilibrium point, or the pressure is relieved (and we stop throttling) or we max throttle, in which case the Operation processor itself will stop processing anything until we relieve pressure.